### PR TITLE
[release-1.6] 🌱 Support MachinePools without MachinePoolMachines in clusterctl upgrade test

### PR DESCRIPTION
**Pull request by willie-yao**
**What this PR does / why we need it**:
This PR is a manual cherry-pick of #10498 and #10584

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/area testing

### CODE REVIEW

File: test/e2e/clusterctl_upgrade.go [hunk 1]
```diff
@@ -435,6 +462,26 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			return n, nil
 		}, input.E2EConfig.GetIntervals(specName, "wait-worker-nodes")...).Should(Equal(*controlPlaneMachineCount+*workerMachineCount), "Timed out waiting for all machines to be exist")
```
[the remark this answers is not in the document]
Comment (willie-yao) [reply test/e2e/clusterctl_upgrade.go]: Whoops, must be a rebase issue!

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve

**Pull request merged**